charger/pulsares.go

```diff
@@ -0,0 +1,161 @@
+package charger
+
+// LICENSE
+
+// Copyright (c) 2023 premultiply
+
+// This module is NOT covered by the MIT license. All rights reserved.
+
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import (
+	"encoding/binary"
+	"fmt"
+	"time"
+
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/util"
+	"github.com/evcc-io/evcc/util/modbus"
+)
+
+// Pulsares charger implementation
+type Pulsares struct {
+	conn *modbus.Connection
+	curr uint16
+}
+
+const (
+	pulsaresRegStatus     = 31
+	pulsaresRegGetCurrent = 45
+	pulsaresRegSetCurrent = 51
+)
+
+func init() {
+	registry.Add("Pulsares", NewPulsaresFromConfig)
+}
+
+// NewPulsaresFromConfig creates a Pulsares charger from generic config
+func NewPulsaresFromConfig(other map[string]interface{}) (api.Charger, error) {
+	cc := struct {
+		modbus.Settings `mapstructure:",squash"`
+		Timeout         time.Duration
+	}{
+		Settings: modbus.Settings{
+			ID: 1,
+		},
+	}
+
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	return NewPulsares(cc.URI, cc.Device, cc.Comset, cc.Baudrate, modbus.ProtocolFromRTU(cc.RTU), cc.ID, cc.Timeout)
+}
+
+// NewPulsares creates Pulsares charger
+func NewPulsares(uri, device, comset string, baudrate int, proto modbus.Protocol, slaveID uint8, timeout time.Duration) (api.Charger, error) {
+	conn, err := modbus.NewConnection(uri, device, comset, baudrate, proto, slaveID)
+	if err != nil {
+		return nil, err
+	}
+
+	if timeout > 0 {
+		conn.Timeout(timeout)
+	}
+
+	log := util.NewLogger("pulsares")
+	conn.Logger(log.TRACE)
+
+	wb := &Pulsares{
+		conn: conn,
+		curr: 6,
+	}
+
+	// get initial state from charger
+	curr, err := wb.getCurrent()
+	if err != nil {
+		return nil, fmt.Errorf("current limit: %w", err)
+	}
+	if curr >= 6 {
+		wb.curr = curr
+	}
+
+	return wb, err
+}
+
+func (wb *Pulsares) setCurrent(current uint16) error {
+	b := make([]byte, 2)
+	binary.BigEndian.PutUint16(b, uint16(current))
+
+	_, err := wb.conn.WriteMultipleRegisters(pulsaresRegSetCurrent, 1, b)
+
+	return err
+}
+
+func (wb *Pulsares) getCurrent() (uint16, error) {
+	b, err := wb.conn.ReadHoldingRegisters(pulsaresRegGetCurrent, 1)
+	if err != nil {
+		return 0, err
+	}
+
+	return binary.BigEndian.Uint16(b) / 1e3, nil
```

Comment:
1e3 fürs lesen?

Reply:
@wb3103 sehr schade, dass wir den Strom nicht in mA setzen können- wäre das möglich?

Reply:
Hallo und Guten Abend,
ich kenn mich mit go nicht aus und, na ja, ich bin Baujahr 1949, werde mich wohl dort nicht mehr einarbeiten können. Es fehlt sozusagen der Lehrer und auch die (verbleibende) Zeit kann knapp werden.
Den Strom kann man im Register 0x5d in mA setzen.
In den Ladeprofilregistern (0x33 bis 0x3b) wird der Strom in A-Schritten geschrieben.
Ich weiß nicht, wie ich die Software compilieren und testen kann, hier brauch ich Hilfe.
MfG Winne

Reply:
> Den Strom kann man im Register 0x5d in mA setzen.

Ok, findet man das irgendwo dokumentiert?

Ich hab die Doku so verstanden dass man Profil 1 ändern muss wenn keine Profile benutzt werden sollen.

Reply:
Richtig, im Register 5D ist der Ladestrom in mA setzbar. Das Profil-System ist eine zusätzliche Limitierung. Wird der Stecker angesteckt, schaltet die Steuerung in Profil 1. Es wird immer der niedrigste Ladestrom geschaltet.

Reply:
Watchdog, dient eigentlich nur der Selbstüberwachung der Box. Siehe Seite 11, 32 und 42 der Betriebsanleitug EV EasyCharge.